### ISSUE TITLE
Could not optimize @tailrec annotated method factorial

### ISSUE THREAD

**Issue by sbushmanov** [scalameta/metals#2907]
When I do in scala worksheet:

```scala
  import scala.annotation.tailrec

  @tailrec
>>def factorial(n: BigInt, acc: BigInt): BigInt = {
    if (n <=1) acc
    else factorial(n-1, n*acc)
  }

  factorial(3,1)

```
I'm getting an error:
```
[mdoc] [E] could not optimize @tailrec annotated method factorial: it is
   neither private nor final so can be overridden
```

Expected behavior:
`factorial` should be properly evaluated without an error.

Setup:
- vim version: NVIM v0.5.0-dev+1444-gc1120ad          
- node version: v15.14.0                                
- coc.nvim version: 0.0.80-131f6b267f 
- same error for both bloop and sbt

**Comment by tgodzik** [comment]
Thanks for reporting! We already were fixing it upstream, but forgot to update the fix in Metals. Should work after #2908 is released

**Comment by sbushmanov** [comment]
Perhaps not related, but may I ask why for a problem diagnostic to be identified I need to reload the build?

**Comment by tgodzik** [comment]
You should not need to reload anything to get diagnostics, in case of worksheets it should work on save. However, if there is an internal error you might need to reload the worksheet or force another save.

**Comment by sbushmanov** [comment]
In `*.scala` files I'm only getting diagnostics on save.  In `*.worksheet.sc` I need to reload build.
Am I missing something? 
Or should I open an issue?

**Comment by tgodzik** [comment]
We only publish diagnostics on save aside from syntax errors, so that is expected.

For worksheets only saving is also needed. How do you reloaf the build? What's the snippet that you are using? Probably best to raise a separate issue with a reproduction
